### PR TITLE
add policy tests for vault in different namespaces

### PR DESCRIPTION
**Pull request by xshl202007**
## Motivation

1) currently the vault's policy-related tests (in secure/storage/src/tests/vault.rs) are done assuming no namespace
2) and due to this, some codes are not in test coverage (like https://codecov.io/gh/libra/libra/src/master/secure/storage/src/vault.rs: create_token(): line 134-137)

the changes here:
1) add some unit tests to verify the policy still works with some namespace specified
2) verify permission error when do cross-namespace operations (i.e. policy won't work cross namespaces)
3) fix some test coverage as mentioned above

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan
The changes here are: added more unit tests
1) "cargo x lint && cargo xfmt && cargo xclippy --all-targets" OK: (as in overall_cargo_test_results.jpg)
2) "cargo test" OK: (as in cmd_cargo_test_results.jpg)

## Related PRs
N/A

cmd_cargo_test_results
overall_cargo_test_results

### CODE REVIEW

**Review by davidiw — COMMENTED** [review]
Thanks for the contribution! Can we shrink this down to just a handful of tests? Adding 250 lines of code to increase test coverage by 4 lines doesn't seem like a great tradeoff.

Also if this area is of interest to you, I can actually funnel you more challenging / interesting work :D

File: secure/storage/src/tests/vault.rs [hunk 1]
```diff
@@ -181,6 +195,90 @@ fn test_vault_key_value_policies() {
...
+    // Verify a reader with another namespace has no permission for the operations
+    let mut reader_with_namespace = VaultStorage::new(
+        dev::test_host(),
+        reader_token,
+        Some(VAULT_NAMESPACE_1.into()),
+        None,
+        ttl,
+        false,
+    );
+    assert_eq!(
```
Comment (JoshLind) [inline secure/storage/src/tests/vault.rs:208]: Perhaps we could just check one of these keys to make sure different namespaces really aren't accessible? Checking of all them seems a little excessive to me.

File: secure/storage/src/tests/vault.rs [hunk 2]
```diff
@@ -181,6 +195,90 @@ fn test_vault_key_value_policies() {
...
+    // Verify a writer with another namespace has no permission for the operations
+    let mut writer_with_namespace = VaultStorage::new(
+        dev::test_host(),
+        writer_token,
+        Some(VAULT_NAMESPACE_1.into()),
+        None,
+        ttl,
+        false,
+    );
+    assert_eq!(
```
Comment (JoshLind) [inline secure/storage/src/tests/vault.rs:250]: Likewise, here: we could probably just check one key?

File: secure/storage/src/tests/vault.rs [hunk 3]
```diff
@@ -210,61 +308,197 @@ fn test_vault_crypto_policies() {
...
+    // Verify a exporter with another namespace has no permission for the operations
+    let mut exporter_store_with_namespace = VaultStorage::new(
+        dev::test_host(),
+        exporter_token,
+        Some(VAULT_NAMESPACE_1.into()),
+        None,
+        None,
+        true,
+    );
+    exporter_store_with_namespace
```
Comment (JoshLind) [inline secure/storage/src/tests/vault.rs:332]: It's probably only worth checking that "export_private_key" didn't work. The rest won't work even with the correct namespace (due to permissions -- see the calls above).

File: secure/storage/src/tests/vault.rs [hunk 4]
```diff
@@ -210,61 +308,197 @@ fn test_vault_crypto_policies() {
...
+    // Verify a noone with another namespace has no permission for the operations
+    let mut noone_store_with_namespace = VaultStorage::new(
+        dev::test_host(),
+        noone_token,
+        Some(VAULT_NAMESPACE_1.into()),
+        None,
+        None,
+        true,
+    );
+    noone_store_with_namespace
```
Comment (JoshLind) [inline secure/storage/src/tests/vault.rs:368]: These checks are probably redundant given that noone_store can't even access the keys. I'd remove these checks.

File: secure/storage/src/tests/vault.rs [hunk 5]
```diff
@@ -210,61 +308,197 @@ fn test_vault_crypto_policies() {
...
+    // Verify a reader with another namespace has no permission for the operations
+    let mut reader_store_with_namespace = VaultStorage::new(
+        dev::test_host(),
+        reader_token,
+        Some(VAULT_NAMESPACE_1.into()),
+        None,
+        None,
+        true,
+    );
+    reader_store_with_namespace
```
Comment (JoshLind) [inline secure/storage/src/tests/vault.rs:407]: I'd just check that "get_public_key" doesn't work here. I'm not sure we gain much by checking all the variants.

File: secure/storage/src/tests/vault.rs [hunk 6]
```diff
@@ -210,61 +308,197 @@ fn test_vault_crypto_policies() {
...
+        rotater_token,
+        Some(VAULT_NAMESPACE_1.into()),
+        None,
+        None,
+        true,
+    );
+    rotater_store_with_namespace
+        .export_private_key(CRYPTO_KEY)
+        .unwrap_err();
+    rotater_store_with_namespace
```
Comment (JoshLind) [inline secure/storage/src/tests/vault.rs:449]: Likewise: just check that "get_public_key" doesn't work. Skip the others.

File: secure/storage/src/tests/vault.rs [hunk 7]
```diff
@@ -210,61 +308,197 @@ fn test_vault_crypto_policies() {
...
+    signer_store_with_namespace
+        .export_private_key(CRYPTO_KEY)
+        .unwrap_err();
+    signer_store_with_namespace
+        .get_public_key(CRYPTO_KEY)
+        .unwrap_err();
+    signer_store_with_namespace
+        .rotate_key(CRYPTO_KEY)
+        .unwrap_err();
+    signer_store_with_namespace
```
Comment (JoshLind) [inline secure/storage/src/tests/vault.rs:495]: Just check for this.

File: secure/storage/src/tests/vault.rs [hunk 8]
```diff
@@ -278,12 +512,31 @@ fn test_vault_tokens() {
...
+        None,
+        None,
+        true,
+    );
 
     // Verify reads and write succeed
     assert_eq!(writer.get::<u64>(PARTIAL).unwrap().value, 3);
     writer.set::<u64>(PARTIAL, 5).unwrap();
 
+    // Verify a writer with another namespace has no permission for the operations
```
Comment (JoshLind) [inline secure/storage/src/tests/vault.rs:528]: nit: change the comment to: "// Verify a writer without a namespace has ..."

**Review by JoshLind — COMMENTED** [review]
Thanks for the PR! 😄 I've added some comments. Overall I think checking this functionality is useful, but I'm a little reluctant to check every possible variant for completeness sake. A subset of tests should be enough to give us some confidence. See my comments below.

**Comment by xshl202007** [comment]
thanks a lot for the help and review! 
seems local branch got some mess up, closing this one, and  have updated the codes based on your comments, at https://github.com/libra/libra/pull/6285
thanks a lot;

**Pull request closed without merging**